### ISSUE TITLE
Router config for async resolved properties

### ISSUE THREAD

**Issue by ramnivas** [japgolly/scalajs-react#68]
I just started using the new scalajs-react router and I am running into an issue.

How do I configure a route to change the view only upon resolving properties needed by that view--if the resolving involves one or more async calls? For example, for the `/person/123` route, I want to first resolve (say by making a REST call that returns a `Furute[Person]`) the `Person` object associated with id `123` and only then render the `PersonComponent` (passing it the resolved `Person` object).

In AngularJS, in the same situation, I would use `resolve` [as described here](http://odetocode.com/blogs/scott/archive/2014/05/20/using-resolve-in-angularjs-routes.aspx)

**Comment by japgolly** [comment]
Have a look at https://github.com/japgolly/scalajs-react/blob/master/extra/ROUTER.md#dynamic-routes

In this example it matches `/person/xxx` and behaves differently depending on whether `xxx` is a valid number or not. You just insert whatever logic you want the router to do in there. You can issue an ajax request and render `<.div("loading...")` until the async callback, or you can do something else, it's up to you.

Basically the router allows you make ajax calls, or anything else you want to do, but it doesn't provide any async helpers.

**Comment by ramnivas** [comment]
Thanks, David. That helped me to have it working. Here is what I did:

```
register(parser {
  case personPathMatch(id) => id
}.thenMatch(id => {
  PersonStore.findPerson(id).andThen {
    case Success(p) => React.render(Menu.wrap(PersonEditor(p)), dom.document.body)
    case Failure(ex) => React.render(Menu.wrap(div("Failed")), dom.document.body)
  }
  render(div("Loading..."))
}))
```

`Menu.wrap` is what I normally do in `interceptRender`. Any way to pipe the component created on future completion such that I don't have to repeat the `interceptRender` logic?

**Comment by japgolly** [comment]
You could expose `interceptReader` in your `RoutingRules` but you cross the line into hacking territory if you do that. There's a better way though, I suggest one of these:

1. Have a component that in its Props, takes an input like `PersonId(123)`, and in its state is has `Option[Either[String, Person]]` which is initially `None`. When the component starts up, it sees its state is `None` so it displays something like `"Loading..."` then sends off the ajax request (be careful to track if you've already sent one). When the ajax returns, the callback then just calls `component.setState(Some(Right(person)))` or `component.setState(Some(Left("You don't have access."))`. That way you're not coupled to the router, which means you work nicely with it too.
2. Similar to above but instead of one component for `Person` with ajax loading support, you abstract the `Person` part out to mean anything. That way you have one component for ajax initialising, handling failure and success, which you can separate from (and wrap) all your other components.

Also it's recommended that instead of storing the state in the component itself you just feed it callbacks through Props which allows you to push your state all the way to the top. Maybe ask the React JS mailing list if you'd like some more detail on that.
